src/compiler/factory.ts

```diff
@@ -2292,15 +2292,26 @@ namespace ts {
     /**
      * Sets a custom text range to use when emitting source maps.
      */
-    export function setSourceMapRange<T extends Node>(node: T, range: TextRange | undefined) {
+    export function setSourceMapRange<T extends Node>(node: T, range: SourceMapRange | undefined) {
         getOrCreateEmitNode(node).sourceMapRange = range;
         return node;
     }
 
     /**
+     * Create an external source map source file reference
+     */
+    export function createSourceMapSource(fileName: string, text: string, skipTrivia?: (pos: number) => number) {

```

Comment:
@chuckjaz The constructor for a `SourceFile` is actually created by the [`objectAllocator`](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/core.ts#L2244). This allows us to keep nodes relatively light for the commandline compiler, but also allows us to create richer nodes for the language service.

You could consider adding a constructor factory for `SourceMapSource` to the object allocator, and hook in the `getLineAndCharacterOfPosition` method with a separate constructor in services.